src/udev/hidraw_id/hidraw_id.c

```diff
@@ -0,0 +1,104 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Identifies HID device types (e.g. security tokens) based on the usages declared in the report descriptor
+ * and outputs suitable environment variables.
+ *
+ * Modified from "u2f_hidraw_id.c" on 2019-08-20
+ *
+ * Original copyright:
+ *
+ * Copyright (c) 2014-2015 Andrew Lutomirski
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <linux/hid.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "device-internal.h"
+#include "device-private.h"
+#include "device-util.h"
+#include "fd-util.h"
+#include "hidraw_id_desc.h"
+#include "log.h"
+#include "macro.h"
+#include "main-func.h"
+#include "path-util.h"
+#include "string-util.h"
+#include "udev-util.h"
+
+static int run(int argc, char **argv) {
+        _cleanup_(sd_device_unrefp) struct sd_device *device = NULL;
+        _cleanup_free_ char *desc_path = NULL;
+        _cleanup_close_ int fd = -1;
+
+        struct sd_device *hid_device;
+        const char *sys_path;
+        uint8_t desc[HID_MAX_DESCRIPTOR_SIZE];
+        ssize_t desc_len;
+        struct HidrawIdDescResult desc_result;
+
+        int r;
+
+        log_set_target(LOG_TARGET_AUTO);
+        udev_parse_config();
+        log_parse_environment();
+        log_open();
+
+        if (argc != 2) {
+                fprintf(stderr, "Usage: hidraw_id SYSFS_PATH|--rule\n");
+                return EXIT_FAILURE;
+        }
+
+        if (streq(argv[1], "--rule")) {
```

Comment:
hmm, why this?

Reply:
The `if` is there so that the program can be run outside the context of a udev rule (e.g. troubleshooting). I renamed the command line flag to `--rule` to indicate that it is supposed to be used within a rule, as I wasn't sure whether `udev` is a good description for its context. Would you say that `--udev` or something else entirely is better?

Reply:
How about the following:
`argc == 1`: use `environ`, 
`argc == 2`: use `argv[1]`.

Reply:
I like this idea a lot and just implemented it.